linker/Linker.Steps/MarkStep.cs

```diff
@@ -1587,5 +1598,116 @@ protected virtual void MarkInterfaceImplementation (TypeDefinition type, Interfa
+				if ((bindingFlags == BindingFlags.Default || bindingFlags.HasFlag (BindingFlags.Public) == method.IsPublic) && method.Name == ".ctor")
+					MarkMethod (method);
+			}
+		}
+
+		void MarkMethodUsedViaReflection (Collection<Instruction> instructions, string nameOfMethodInvokedViaReflection, TypeDefinition declaringTypeOfMethodInvokedViaReflection, BindingFlags bindingFlags)
+		{
+			if (nameOfMethodInvokedViaReflection != null) {
+				foreach (var method in declaringTypeOfMethodInvokedViaReflection.Methods)
+				{
+					if ((bindingFlags == BindingFlags.Default || bindingFlags.HasFlag (BindingFlags.Public) == method.IsPublic && bindingFlags.HasFlag(BindingFlags.Static) == method.IsStatic)
+						&& method.Name == nameOfMethodInvokedViaReflection)
+						MarkMethod (method);
+				}
+			}
+		}
+
+		void MarkPropertyUsedViaReflection (Collection<Instruction> instructions, string nameOfPropertyInvokedViaReflection, TypeDefinition declaringTypeOfPropertyInvokedViaReflection, BindingFlags unused)
+		{
+			if (nameOfPropertyInvokedViaReflection != null) {
+				foreach (var property in declaringTypeOfPropertyInvokedViaReflection.Properties) {
+					if (property.Name == nameOfPropertyInvokedViaReflection) {
+						// It is not easy to reliably detect in the IL code whether the getter or setter (or both) are used.
+						// Be conservative and mark everything for the property.
+						MarkProperty (property);
+						MarkMethodIfNotNull (property.GetMethod);
+						MarkMethodIfNotNull (property.SetMethod);
+					}
+				}
+			}
+		}
+
+		void MarkFieldUsedViaReflection (Collection<Instruction> instructions, string nameOfFieldInvokedViaReflection, TypeDefinition declaringTypeOfFieldInvokedViaReflection, BindingFlags unused)
+		{
+			if (nameOfFieldInvokedViaReflection != null) {
+				foreach (var field in declaringTypeOfFieldInvokedViaReflection.Fields) {
+					if (field.Name == nameOfFieldInvokedViaReflection)
+						MarkField (field);
+				}
+			}
+		}
+
+		void MarkEventUsedViaReflection (Collection<Instruction> instructions, string nameOfEventInvokedViaReflection, TypeDefinition declaringTypeOfEventInvokedViaReflection, BindingFlags unused)
+		{
+			if (nameOfEventInvokedViaReflection != null) {
+				foreach (var eventInfo in declaringTypeOfEventInvokedViaReflection.Events) {
+					if (eventInfo.Name == nameOfEventInvokedViaReflection)
+						MarkEvent (eventInfo);
+				}
+			}
+		}
+
+		static TOperand OperandOfNearestInstructionBefore<TOperand> (int startingInstructionIndex, OpCode opCode, IList<Instruction> instructions)
+		{
+			for (var i = startingInstructionIndex; i >= 0; i--) {
+				if (instructions [i].OpCode == opCode)
+					return (TOperand )instructions [i].Operand;
+			}
+
+			return default (TOperand);
+		}
+
+		static List<TOperand> OperandsOfInstructionsBefore<TOperand> (int startingInstructionIndex, OpCode opCode, IList<Instruction> instructions) where TOperand : class
```

Comment:
I think this is unnecessary generalization, it should check for ldtoken with TypeReference

Reply:
I agree, I'll remove the generic here.